gitlab/resource_gitlab_project_protected_environment.go

```diff
@@ -0,0 +1,215 @@
+			d.SetId("")
+			return nil
+		}
+		return fmt.Errorf("error getting gitlab project %q protected environment %q: %v", project, environment, err)
+	}
+
+	d.Set("environment", protectedEnvironment.Name)
+	if err := d.Set("deploy_access_levels", flattenDeployAccessLevels(protectedEnvironment.DeployAccessLevels)); err != nil {
+		return fmt.Errorf("error setting deploy_access_levels: %v", err)
+	}
+
+	return nil
+}
+
+func resourceGitlabProjectProtectedEnvironmentDelete(d *schema.ResourceData, meta interface{}) error {
+	project, environmentName, err := parseTwoPartID(d.Id())
+	if err != nil {
+		return err
+	}
+
+	log.Printf("[DEBUG] Project %s delete gitlab project-level protected environment %s", project, environmentName)
+
+	client := meta.(*gitlab.Client)
+
+	_, err = client.ProtectedEnvironments.UnprotectEnvironment(project, environmentName)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func expandDeployAccessLevels(vs []interface{}) ([]*gitlab.EnvironmentAccessOptions, error) {
+	result := make([]*gitlab.EnvironmentAccessOptions, 0)
+
+	for _, v := range vs {
+		opts := v.(map[string]interface{})
+		option := &gitlab.EnvironmentAccessOptions{}
+		if accessLevel, exists := opts["access_level"]; exists {
+			accessLevelValue, err := accessLevelStringToValue(accessLevel.(string))
+			if err != nil {
+				return nil, fmt.Errorf("error expanding access level: %v", err)
+			}
+			option.AccessLevel = &accessLevelValue
+		} else if userID, exists := opts["user_id"]; exists {
+			option.UserID = gitlab.Int(userID.(int))
+		} else if groupID, exists := opts["group_id"]; exists {
+			option.GroupID = gitlab.Int(groupID.(int))
+		}
+		result = append(result, option)
+	}
+
+	return result, nil
+}
+
+func flattenDeployAccessLevels(vs []*gitlab.EnvironmentAccessDescription) []map[string]interface{} {
+	result := make([]map[string]interface{}, 0)
+
+	for _, accessDescription := range vs {
+		v := make(map[string]interface{})
+		v["access_level"] = accessLevelValueToName[accessDescription.AccessLevel]
+		v["access_level_description"] = accessDescription.AccessLevelDescription
+		if accessDescription.UserID != 0 {
+			v["user_id"] = accessDescription.UserID
+		}
+		if accessDescription.GroupID != 0 {
+			v["group_id"] = accessDescription.GroupID
+		}
+		result = append(result, v)
+	}
+
+	return result
+}
+
+func accessLevelStringToValue(level string) (gitlab.AccessLevelValue, error) {
+	switch level {
+	case "noone":
+		return gitlab.NoPermissions, nil
+	case "minimal":
+		return gitlab.MinimalAccessPermissions, nil
+	case "guest":
+		return gitlab.GuestPermissions, nil
+	case "reporter":
+		return gitlab.ReporterPermissions, nil
+	case "developer":
+		return gitlab.DeveloperPermissions, nil
+	case "maintainer":
+		return gitlab.MaintainerPermissions, nil
+	case "owner":
+		return gitlab.OwnerPermissions, nil
+	default:
+		return -1, fmt.Errorf("unknown access level %q", level)
+	}
+}
```

Comment:
This seems to be duplicating `accessLevelNameToValue` (https://github.com/gitlabhq/terraform-provider-gitlab/blob/master/gitlab/util.go#L15), but with some differences (is `noone` correct, or `no one`?).

Would it make sense to fix/update `accessLevelNameToValue` instead?

Reply:
Yes, certainly. I'll make the change, remove the above code and replace uses of it with `accessLevelNameToValue`.